File: backend/users/permissions.py
"""Permisos DRF basados en el rol del usuario.

La política se define aquí (una sola fuente de verdad) y las vistas la aplican.
El frontend oculta/deshabilita botones, pero la autorización real vive en el
backend: aunque alguien llame la API a mano, estos permisos la protegen.
"""
from __future__ import annotations

from rest_framework.permissions import BasePermission


class IsAdminRole(BasePermission):
    """Solo administradores (rol admin o superusuario)."""

    message = 'Se requiere rol de Administrador.'

    def has_permission(self, request, view) -> bool:
        user = request.user
        return bool(user and user.is_authenticated and user.is_admin_role)


class IsSuperAdmin(BasePermission):
    """Solo el administrador general (superusuario), el que ve todas las empresas."""

    message = 'Se requiere ser Administrador general.'

    def has_permission(self, request, view) -> bool:
        user = request.user
        return bool(user and user.is_authenticated and user.is_superadmin)


class CanOperate(BasePermission):
    """Solo operadores o administradores (acciones de escritura/gestión).

    Excluye al administrador general (ver `User.can_operate`): es de solo lectura
    sobre los datos de las empresas.
    """

    message = 'Se requiere rol de Operador o Administrador.'

    def has_permission(self, request, view) -> bool:
        user = request.user
        return bool(user and user.is_authenticated and user.can_operate)


class IsNotGlobalAdmin(BasePermission):
    """Permite a cualquier usuario de empresa (incluido Consulta) pero NO al
    administrador general.

    Es para acciones que el rol Consulta sí puede hacer pero que mutan datos
    —entregar un Código Pin lo marca como usado y desbloquea el equipo—, de las
    que el auditor global debe quedar fuera para ser de solo lectura de verdad.
    """

    message = 'El administrador general es de solo lectura sobre los televisores.'

    def has_permission(self, request, view) -> bool:
        user = request.user
        return bool(user and user.is_authenticated and not user.is_superadmin)

File: frontend/src/features/auth/permissions.ts
// Política de permisos del frontend (espejo de `users/permissions.py` en el
// backend). Aquí solo se decide qué se MUESTRA; la autorización real la impone
// el backend. Mantener ambas listas alineadas.

import type { Role, User } from '@/features/auth/types'

/**
 * Administrador general: el único que ve todas las empresas y puede crearlas.
 * Los demás (incluido el rol `admin`) están confinados a la suya.
 */
export function isSuperAdmin(user: User | null): boolean {
  return user?.is_superadmin === true
}

/** Administrador: gestión de usuarios + todos los módulos (dentro de su empresa). */
export function isAdmin(user: User | null): boolean {
  return user?.role === 'admin' || isSuperAdmin(user)
}

/**
 * Puede realizar gestiones de escritura sobre los datos (crear/editar TV,
 * cambiar estados, enrolar, sincronizar).
 *
 * El administrador general queda FUERA: es un auditor de solo lectura sobre los
 * datos de las empresas (los ve y exporta todos, pero no los altera). Su poder
 * está en otro plano —usuarios y empresas— que no pasa por aquí. Espejo de
 * `User.can_operate` en el backend.
 */
export function canOperate(user: User | null): boolean {
  if (isSuperAdmin(user)) return false
  return user?.role === 'operador' || user?.role === 'admin'
}

export const ROLE_LABELS: Record<Role, string> = {
  admin: 'Administrador',
  operador: 'Operador',
  consulta: 'Consulta',
}
